utils/v2/connector/impl/RedStonePriceFeed.ts

```diff
@@ -1,47 +1,116 @@
+
+export interface DataSourcesConfig {
+  valueSelectionAlgorithm: ValueSelectionAlgorithm,
+  timeoutMilliseconds: number;
+  maxTimestampDiffMilliseconds: number;
+  preVerifySignatureOffchain: boolean;
+  signers?: string[], // Will be used in future (after multi-nodes creation)
+  sources: SourceConfig[],
+};
+
+export interface PriceFeedOptions {
+  dataSources?: DataSourcesConfig;
+  asset?: string;
+};
+
+export type RedStoneProvider =
+  | "redstone"
+  | "redstone-stocks"
+  | "redstone-rapid"
+  | "redstone-avalanche";
 
 export class RedStonePriceFeed implements PriceFeedConnector {
 
-  private readonly priceSigner = new EvmPriceSigner();
-  private readonly apiUrl: string;
+  // TODO: remove
+  // private readonly priceSigner = new EvmPriceSigner();
   private cachedSigner?: string;
+  private fetchers: Fetcher[] = [];
 
   constructor(
     private providerId: RedStoneProvider,
-    private asset?: string) {
+    private priceFeedOptions: PriceFeedOptions = {}) {
 
-    this.apiUrl = `https://api.redstone.finance/packages/latest?provider=${providerId}`
-      + (asset ? `&symbol=${asset}` : '');
-    
+      // Getting default data sources config for provider if not specified
+      if (!this.priceFeedOptions.dataSources) {
+        this.priceFeedOptions.dataSources = getDefaultDataSourceConfig(providerId);
+      }
+
+      // Init fetchers
+      for (const [i, source] of this.priceFeedOptions.dataSources!.sources!.entries()) {
+        if (priceFeedOptions.asset && source.disabledForSinglePrices) {
+          console.log(`Skipping ${i} (${source.type}) source init`);
+        } else {
+          const fetcherForSource = createFetcher(source, priceFeedOptions.asset);
+          this.fetchers.push(fetcherForSource);
+        }
+      }
+
+      // TODO: get rid of it, it's a potential single point of failure
+      this.getSigner(); // we are loading signer public key in advance
   }
 
+  // This is the entrypoint function of this module
   async getSignedPrice(): Promise<SignedPriceDataType> {
-    const response = await axios.get(this.apiUrl);
+    // We need to get signer public key firstly
+    // It will be used to pre-verify signatures off-chain
+    await this.getSigner();
+    
+    // Fetching data from all sources simultaneously with timeout
+    const timeoutMilliseconds =
+      this.priceFeedOptions.dataSources?.timeoutMilliseconds
+      || DEFAULT_TIMEOUT_MILLISECONDS;
+    const promises = this.fetchers.map(fetcher =>
+      fetcher.getLatestDataWithTimeout(this.providerId, timeoutMilliseconds));
+    const results = await Promise.allSettled(promises);
+
+    // Validating fetched data
+    const fulfilledPromisesResults = results.filter(r => r.status === "fulfilled") as
+      PromiseFulfilledResult<SignedDataPackageResponse>[];
+    const dataPackages = fulfilledPromisesResults.map(r => r.value);
+    const validDataPackages = dataPackages.filter(p =>
```

Comment:
How expensive is validation?
An alternative approach will be to first optimistically select the package (latest) and validate only a single package.

Reply:
I think, it's not really expensive, but will do performance testing

Reply:
Hmm, I think it's more important to implement "first-valid" approach first. 
I'm really afraid to be bound by the worst/slowest fetcher every time.

Reply:
Ok, I will implement it

Reply:
It's completed